### DISCUSSION TITLE
How to make BackgroundTasks manager?

### DISCUSSION THREAD

**Discussion by fookashi** [fastapi/fastapi#9948] [Questions] [answered]
### Description

Every request i do, it appears new object of BackgroundTasks and tasks are completing parallel, so it`s make motor(async driver for mongodb) to upload csvs in parallel(and instead of to upload one by one csvs for 8 seconds for each, it is uploading 3 csvs together for 30 seconds). I don't know what to do exactly to make requests in queue and not to block endpoints!!!

### Stack
FastAPI,Motor, polars

### Example Code

```python
#Router for uploading csv
@UploadRouter.post(
    "/",
    response_model=CsvUploadResponse,
    status_code=status.HTTP_201_CREATED
)
async def upload(
    bg:BackgroundTasks,
    file : UploadFile = File(...),
    uploadService : UploadService = Depends()
):
    return uploadService.add_to_queue(file=file,bg=bg)

#service method for adding csv to queue
def add_to_queue(self, file: File, bg):
    content_type = file.content_type
    if content_type != "text/csv":
        raise raise_with_log(status_code=415,detail='Only csv files are supported!')
    name = uuid4().hex
    self.statusRepository.create(name)
    bg.add_task(self.upload,file,name)
    return CsvUploadResponse(status=True, msg='Csv is in uploading queue now', id=name)

#service method for uploading csv to mongodb collection
async def upload(self,file,name):
    t1=time.time()
    self.statusRepository.update(name,'Uploading')
    temp_file = BytesIO()
    while chunk:=await file.read(10*1024*1024):
        temp_file.write(chunk)
    try:
        csv=polars.read_csv(temp_file).to_dict()
    except polars.ComputeError:
        self.statusRepository.update(name,'Failed')
        return
    csv_data = (dict(zip(csv.keys(), row)) for row in zip(*csv.values()))
    await self.csvRepository.create(csv_data,name)
    self.statusRepository.update(name,'Uploaded')
    t2=time.time()
    print('ALL TIME IS: ',t2-t1)
    
#csv repository method for inserting csv in collection
async def create(self,data,csv_name):
    collection = self._db[csv_name]
    t1=time.time()
    await collection.insert_many(data)
    t2=time.time()
    print('INSERT TIME: ',t2-t1)
```

### Operating System

macOS

### FastAPI Version

0.100.0

### Python Version

3.11

**Comment by arjwilliams** [comment, marked as answer]
Hi @fookashi,

There are a couple of things:

1. You are calling heavy synchronous code from inside your asynchronous upload function. Therefore, this will block the asyncio event loop, so no new requests will be handled until it finishes. You will need to use an [thread pool executor](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor) (as I believe polars releases the GIL - if not, you will need to use a process pool executor) to be able to run the synchronous code in a non-blocking way. FastAPI provides a helper function 'run_in_threadpool' that uses a default thread pool, or you can look into https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor.
2. You don't need to create the temp_file object - FastAPI's UploadFile is just an asynchronous wrapper around a standard synchronous temporary file, and you can simply use that synchronous file, rather than copying things over (see https://fastapi.tiangolo.com/tutorial/request-files/?h=uploadfile#uploadfile).
3. With the fixes from part 1, I'm not sure whether you will still want to run the upload function one at a time. If so, you can just use a lock (https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock).

So your code would look something like:

```python
import asyncio
from fastapi.concurrency import run_in_threadpool

#Router for uploading csv
@UploadRouter.post(
    "/",
    response_model=CsvUploadResponse,
    status_code=status.HTTP_201_CREATED
)
async def upload(
    bg:BackgroundTasks,
    file : UploadFile = File(...),
    uploadService : UploadService = Depends()
):
    return await uploadService.add_to_queue(file=file,bg=bg)

#service method for adding csv to queue
async def add_to_queue(self, file: File, bg):
    content_type = file.content_type
    if content_type != "text/csv":
        raise raise_with_log(status_code=415,detail='Only csv files are supported!')
    name = uuid4().hex
    await run_in_threadpool(self.statusRepository.create, name)
    bg.add_task(self.upload,file,name)
    return CsvUploadResponse(status=True, msg='Csv is in uploading queue now', id=name)

# define lock somewhere in service
self.lock = asyncio.Lock()

#service method for uploading csv to mongodb collection
async def upload(self,file,name):
    t1=time.time()
    with self.lock:
      await run_in_threadpool(self.statusRepository.update, name, 'Uploading')
      try:
          csv = await run_in_threadpool(polars.read_csv, file.file)
          csv_data = await run_in_threadpool(csv.to_dict)
      except polars.ComputeError:
          await run_in_threadpool(self.statusRepository.update, name, 'Failed')
          return
      csv_rows = (dict(zip(csv_data.keys(), row)) for row in zip(*csv_data.values()))
      await self.csvRepository.create(csv_rows,name)
      await run_in_threadpool(self.statusRepository.update, name, 'Uploaded')
    t2=time.time()
    print('ALL TIME IS: ',t2-t1)
```

**Reply by fookashi** [reply]
Hello, thank you for your answer. Yes, you are correct, but to make it works in more sync way(because when i'm init asyncio.Lock() in service, it doesn't change anything, so i did something like this:
**In main.py**
`app=create_app()`
`lock=asyncio.Lock()`
**In UploadService**
`from main import lock`
`async with lock:`

And if i do some size validation, i can use both: parallel and not parallel inserting.For example, if size<5MB then upload in parallel(without lock), else add to queue!
But i still have some questions:
1) Is it right to use so many _run_in_threadpool_? I read that better way to use processpoolexecutor, because a lot of threads will get app stuck.
2) Why do i need BackgroundTasks, if i don`t use its queue? If i send 10 requests, there would be 10 different BackgroundTasks, but how to fix it?

**Reply by arjwilliams** [reply]
1. In asynchronous code, there always needs to be some way to pass control back to the asyncio event loop, otherwise the function will essentially run synchronously (it will block other asynchronous functions from running). The run_in_threadpool does that for you by wrapping a thread pool in an asyncronous wrapper. You can use the process pool executor instead, but then you will need to use `await asyncio.get_running_loop().run_in_executor(processPoolExecutor, ...)`, as just using the standard synchronous futures/promises from the process pool will block.
I'm not sure whether the process pool or the thread pool will be better in your case. You might want to benchmark both cases.
You can also refactor things, moving the synchronous code into different functions that will make it slightly better:
```python
def read_csv(file: UploadFile):
    return polars.read_csv(file.file).to_dict()

async def updateStatus(name: str, status: str):
    await run_in_threadpool(self.statusRepository.update, name, status)

async def upload(self,file,name):
    t1=time.time()
    async with self.lock:
      await updateStatus(name, 'Uploading')
      try:
          csv_data = await run_in_threadpool(read_csv, file)
      except polars.ComputeError:
          await updateStatus(name, 'Failed')
          return
      csv_rows = (dict(zip(csv_data.keys(), row)) for row in zip(*csv_data.values()))
      await self.csvRepository.create(csv_rows,name)
      await updateStatus(name, 'Uploaded')
    t2=time.time()
    print('ALL TIME IS: ',t2-t1)
```
2. That is expected behaviour. If the same BackgroundTasks instance was returned each time, then the old tasks would be awaited on multiple times (the BackgroundTasks class is controlled by Starlette - the parent project of FastAPI - rather than FastAPI itself).
The BackgroundTasks class is super simple - just a list of functions that get called/awaited on after the response has been sent.
What are you expecting to do if you have the same instance?

**Reply by fookashi** [reply]
Hello, [arjwilliams](https://github.com/arjwilliams)! Thank you for answering me again. I started to understand FastAPI processes better because of your replies! 
1) I guess it's better to now init ThreadExecutor, at least i found discussion
 [#3316]
<img width="646" alt="image" src="https://github.com/tiangolo/fastapi/assets/44700352/624223f0-3843-4748-b4f6-3d16b51d46d4">

**EDIT: Noticed that answer was about _def_ functions**

So now, i suppose there can be something like ThreadGuard class, which will execute functions in ThreadPoolExecutor or Semaphore(?):
```
class ThreadGuard:
    def __init__(self, max_threads):
        self.semaphore = anyio.Semaphore(max_threads)
        
    async def run_with_guard(self,func: Callable[P, T], *args: P.args, **kwargs: P.kwargs):
        async with self.semaphore:
            return await anyio.to_thread.run_sync(func, *args, **kwargs)
```

2) I thought i can do instance of BackgroundTasks that would do uploads in queue. I mean, i do 10 requests for uploading, and they will be in queue [1 request, 2 request, 3....]. If i had several background tasks in one request, it would work fine:
``` bg=BackgroundTasks()
bg.add_task(task1)
bg.add_task(task2)
#first complete task1, then task2
```
But if i send 10 requests, task1 and task2 for each requests will complete together. It sounds fine untill situation is about to upload something. If there is hundred requests to upload 500MB files, it would be fine if person who sent requests firstly will get result firstly. But in my case all files are uploaded together. asyncio.Lock() solves this problem, but it seems like there is a smarter approach to the solution. I tried to init BackgroundTasks on event startup, but it seems it doesn't work like that.
